File: benchmarks/suite/spectral_norm/bench.py
import math

def eval_A(i, j):
    return 1.0 / ((i + j) * (i + j + 1) // 2 + i + 1)

def eval_A_times_u(u, n, au):
    for i in range(n):
        s = 0.0
        for j in range(n):
            s += eval_A(i, j) * u[j]
        au[i] = s

def eval_At_times_u(u, n, atu):
    for i in range(n):
        s = 0.0
        for j in range(n):
            s += eval_A(j, i) * u[j]
        atu[i] = s

def eval_AtA_times_u(u, n, atau):
    v = [0.0] * n
    eval_A_times_u(u, n, v)
    eval_At_times_u(v, n, atau)

def main():
    n = 1000
    u = [1.0] * n
    v = [0.0] * n

    for _ in range(10):
        eval_AtA_times_u(u, n, v)
        eval_AtA_times_u(v, n, u)

    vBv = 0.0
    vv = 0.0
    for i in range(n):
        vBv += u[i] * v[i]
        vv += v[i] * v[i]

    print(f"Spectral norm: {math.sqrt(vBv / vv):.9f}")

main()

File: benchmarks/suite/spectral_norm/bench.go
package main

import (
	"fmt"
	"math"
)

func evalA(i, j int) float64 {
	return 1.0 / float64((i+j)*(i+j+1)/2+i+1)
}

func evalATimesU(u []float64, n int, au []float64) {
	for i := 0; i < n; i++ {
		sum := 0.0
		for j := 0; j < n; j++ {
			sum += evalA(i, j) * u[j]
		}
		au[i] = sum
	}
}

func evalAtTimesU(u []float64, n int, atu []float64) {
	for i := 0; i < n; i++ {
		sum := 0.0
		for j := 0; j < n; j++ {
			sum += evalA(j, i) * u[j]
		}
		atu[i] = sum
	}
}

func evalAtATimesU(u []float64, n int, atau []float64) {
	v := make([]float64, n)
	evalATimesU(u, n, v)
	evalAtTimesU(v, n, atau)
}

func main() {
	n := 1000

	u := make([]float64, n)
	v := make([]float64, n)
	for i := range u {
		u[i] = 1.0
	}

	for i := 0; i < 10; i++ {
		evalAtATimesU(u, n, v)
		evalAtATimesU(v, n, u)
	}

	vBv := 0.0
	vv := 0.0
	for i := 0; i < n; i++ {
		vBv += u[i] * v[i]
		vv += v[i] * v[i]
	}

	fmt.Printf("Spectral norm: %.9f\n", math.Sqrt(vBv/vv))
}
